pkg/virt-controller/watch/vm.go

```diff
@@ -428,41 +429,139 @@ func (c *VMController) handleDataVolumes(vm *virtv1.VirtualMachine, dataVolumes
+	runStrategy := vm.RunStrategy()
+	log.Log.Object(vm).V(4).Infof("VirtualMachine RunStrategy: %s", runStrategy)
 
-	if vm.Spec.Running == true {
+	switch runStrategy {
+	case virtv1.RunStrategyRunning:
+		// For this RunStrategy, a VMI should always be running. If a StateChangeRequest
+		// asks to stop a VMI, a new one must be immediately re-started.
 		if vmi != nil {
-			if vmi.IsFinal() {
-				// The VirtualMachineInstance can fail od be finished. The job of this controller
+			forceRestart := false
+			if vm.Status.StateChangeRequest != nil &&
+				vm.Status.StateChangeRequest.StopVMI != nil &&
+				*vm.Status.StateChangeRequest.StopVMI == vmi.UID {
+				forceRestart = true
+			}
+			if forceRestart || vmi.IsFinal() {
+				// The VirtualMachineInstance can fail or be finished. The job of this controller
 				// is keep the VirtualMachineInstance running, therefore it restarts it.
 				// restarting VirtualMachineInstance by stopping it and letting it start in next step
 				err := c.stopVMI(vm, vmi)
 				if err != nil {
 					log.Log.Object(vm).Error("Cannot restart VirtualMachineInstance, the VirtualMachineInstance cannot be deleted.")
-					return err
+					return false, err
 				}
 				// return to let the controller pick up the expected deletion
 			}
 			// VirtualMachineInstance is OK no need to do anything
-			return nil
+			return false, nil
 		}
 
 		err := c.startVMI(vm)
-		return err
-	}
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+
+	case virtv1.RunStrategyRestartOnError:
+		// For this RunStrategy, a VMI should only be restarted if it failed.
+		// If a VMI enters the Succeeded phase, it should not be restarted.
+		if vmi != nil {
+			forceStop := false
+			// If there's a stop request that matches the existing VMI's UUID
+			if vm.Status.StateChangeRequest != nil &&
+				vm.Status.StateChangeRequest.StopVMI != nil &&
+				*vm.Status.StateChangeRequest.StopVMI == vmi.UID {
+				forceStop = true
+			}
+
+			if forceStop || vmi.IsFinal() {
+				// For RestartOnError, this controller should only restart the VirtualMachineInstance
+				// if it failed.
+				err := c.stopVMI(vm, vmi)
+				if err != nil {
+					log.Log.Object(vm).Error("Cannot restart VirtualMachineInstance, the VirtualMachineInstance cannot be deleted.")
+					return false, err
+				}
+				// return to let the controller pick up the expected deletion
+			}
+			// VirtualMachineInstance is OK no need to do anything
+			return false, nil
+		}
+
+		startVMI := false
+		if vm.Status.StateChangeRequest != nil {
+			startVMI = vm.Status.StateChangeRequest.StartVMI
+		}
+
+		// Initially the Phase of the VM will be VmPhaseUnset
+		// Otherwise the Phase will match the previously running VMI's (final) phase
+		// Either of the above cases should cause a VMI to be started, but the Succeeded phase should not
+		if vm.Status.Phase == virtv1.VmPhaseUnset || vm.Status.Phase == virtv1.Failed {
```

Comment:
We shouldn't be tracking the VM's phase as the VMI's phase. These are separate concepts with separate lifecycles. 

For example, it would be strange to see a VM object go from Phase=Succeeded to Phase=Failed or vice versa... That's the kind of thing that happens if we map the underlying VMI's phase to the VM's phase.   

We could use a condition to reflect the VMI's lifecycle on the VM if we wanted though. That could make sense. Even then though, the status isn't something that should be used to determine what action should be taken on an object. The status is just the observed state.